grove/grove.go

```diff
@@ -0,0 +1,72 @@
+/*
+Package grove implements an on-disk storage format for arbor forest
+nodes. This hierarchical storage format is called a "grove", and
+the management type implemented by this package satisfies the
+forest.Store interface.
+
+Note: this package is not yet complete.
+*/
+package grove
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+// FS represents a type that acts as a filesystem. It can create and
+// open files at specific paths
+type FS interface {
+	Open(path string) (*os.File, error)
+	Create(path string) (*os.File, error)
+	OpenFile(path string, flag int, perm os.FileMode) (*os.File, error)
+}
+
+// RelativeFS is a file system that acts relative to a specific path
+type RelativeFS struct {
+	Root string
+}
+
+// ensure RelativeFS satisfies the FS interface
+var _ FS = RelativeFS{}
+
+func (r RelativeFS) resolve(path string) string {
+	return filepath.Join(r.Root, path)
+}
+
+// Open opens the given path as an absolute path relative to the root
+// of the RelativeFS
+func (r RelativeFS) Open(path string) (*os.File, error) {
+	return os.Open(r.resolve(path))
+}
+
+// Create makes the given path as an absolute path relative to the root
+// of the RelativeFS
+func (r RelativeFS) Create(path string) (*os.File, error) {
+	return os.Create(r.resolve(path))
+}
+
+// OpenFile opens the given path as an absolute path relative to the root
+// of the RelativeFS
+func (r RelativeFS) OpenFile(path string, flag int, perm os.FileMode) (*os.File, error) {
+	return os.OpenFile(r.resolve(path), flag, perm)
+}
+
+// Grove is an on-disk store for arbor forest nodes.
+type Grove struct {
```

Comment:
It looks like `Grove` is intended to be an adapter for `RelativeFS`, but a lot of that functionality isn't built yet. Is that right?

Reply:
`Grove` will use `RelativeFS` internally to access on-disk entities. The reason that `RelativeFS` and the `FS` type exist is to allow for easy mocking during test cases. I don't *have* to export the `RelativeFS` type. Do you think I shouldn't?